src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs

```diff
@@ -0,0 +1,646 @@
+            TarHelpers.ReadOrThrow(archiveStream, checksumBytes);
+
+            // Empty checksum means this is an invalid (all blank) entry, finish early
+            if (TarHelpers.IsAllNullBytes(checksumBytes))
+            {
+                return false;
+            }
+
+            TarHelpers.ReadOrThrow(archiveStream, typeFlagByte);
+            TarHelpers.ReadOrThrow(archiveStream, linkNameBytes);
+
+            // Convert the byte arrays
+            _name = TarHelpers.GetTrimmedUtf8String(nameBytes);
+            _mode = TarHelpers.GetTenBaseNumberFromOctalAsciiChars(modeBytes);
+            _uid = TarHelpers.GetTenBaseNumberFromOctalAsciiChars(uidBytes);
+            _gid = TarHelpers.GetTenBaseNumberFromOctalAsciiChars(gidBytes);
+            _size = TarHelpers.GetTenBaseNumberFromOctalAsciiChars(sizeBytes);
+
+            if (_size < 0)
+            {
+                throw new FormatException(string.Format(SR.TarSizeFieldNegative, _name));
+            }
+
+            int mTime = TarHelpers.GetTenBaseNumberFromOctalAsciiChars(mTimeBytes);
+            _mTime = TarHelpers.GetDateTimeFromSecondsSinceEpoch(mTime);
+
+            _checksum = TarHelpers.GetTenBaseNumberFromOctalAsciiChars(checksumBytes);
+
+            // Zero checksum means the whole header is empty
+            if (_checksum == 0)
+            {
+                return false;
+            }
+
+            _typeFlag = (TarEntryType)typeFlagByte[0];
+
+            _linkName = TarHelpers.GetTrimmedUtf8String(linkNameBytes);
+
+            if (_format == TarFormat.Unknown)
+            {
+                _format = _typeFlag switch
+                {
+                    TarEntryType.ExtendedAttributes or
+                    TarEntryType.GlobalExtendedAttributes => TarFormat.Pax,
+
+                    TarEntryType.DirectoryList or
+                    TarEntryType.LongLink or
+                    TarEntryType.LongPath or
+                    TarEntryType.MultiVolume or
+                    TarEntryType.RenamedOrSymlinked or
+                    TarEntryType.SparseFile or
+                    TarEntryType.TapeVolume => TarFormat.Gnu,
+
+                    // V7 is the only one that uses 'V7RegularFile'.
+                    TarEntryType.V7RegularFile => TarFormat.V7,
+
+                    // We can quickly determine the *minimum* possible format if the entry type
+                    // is the POSIX 'RegularFile', although later we could upgrade it to PAX or GNU
+                    _ => (_typeFlag == TarEntryType.RegularFile) ? TarFormat.Ustar : TarFormat.V7
+                };
+            }
+
+            return true;
+        }
+
+        // Reads fields only found in ustar format or above and converts them to their expected data type.
+        // Throws if end of stream is reached or if any conversion fails.
+        private void ReadMagicAttribute(Stream archiveStream)
+        {
+            byte[] magicBytes = new byte[FieldLengths.Magic];
```

Comment:
Same question.  Can this be a stackalloc'd span? (I don't know how big Magic is)

Reply:
Fixed with the rented buffer.